### PR TITLE
add data basa

### PR DESCRIPTION
**Pull request by Rick1Marques**

### CODE REVIEW

File: db/import.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+function random(max, min) {
+  return Math.floor(Math.random() * (max - min) + min);
+}
+
+// sampleSize
+function getStarshipId(arr) {
+  return arr[random(arr.length - 1, 0)]._id;
+}
+
+console.log("Connecting...");
```
Comment (miriam-ertl) [inline db/import.js:18]: Just for clean code reasons, you could refactor the code and delete the console.log later.

`console.log("Connecting...");`

File: db/import.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+await connect(process.env.MONGODB_URI);
+
+console.log("Dropping existing collections...");
```
Comment (miriam-ertl) [inline db/import.js:21]: Just for clean code reasons, you could refactor the code and delete the console.log later.

File: db/import.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+await Listing.collection.drop();
+await Starship.collection.drop();
+await User.collection.drop();
+
+const createdStarships = await Starship.insertMany(starships);
+
+await Promise.all(
+  users.map(async ({ listings, ...user }, index) => {
+    console.log(`Creating user: ${index}...`);
```
Comment (miriam-ertl) [inline db/import.js:30]: Just for clean code reasons, you could refactor the code and delete the console.log later.

File: db/import.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    const newUser = await User.create(user);
+
+    console.log(`Saving listings for user: ${index}...`);
```
Comment (miriam-ertl) [inline db/import.js:33]: Just for clean code reasons, you could refactor the code and delete the console.log later.

File: db/import.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        newUser.starships.push();
+      }
+      newUser.starships.push(starhip_id);
+    }
+
+    await newUser.save();
+  })
+);
+
+console.log("Disconnecting...");
```
Comment (miriam-ertl) [inline db/import.js:59]: Just for clean code reasons, you could refactor the code and delete the console.log later.

**Comment by miriam-ertl** [comment]
👍🏻 Only some clean code snippits could be optimized.

**Comment by moonwave99** [comment]
A couple of comments:

- don't mix functional and imperative programming (i.e. for loops and map);
- don't duplicate utility functions in multiple place (e.g. the `random`) function - create a `utils.js` file for that.

The intuition you had for generating the listings is right! You can write a `sample` function for that - putting it all together in the utils:

```js
// db/utils.js
export function sampleMany(array = [], quantity = 1) {
  // SEE: https://stackoverflow.com/questions/2450954/how-to-randomize-shuffle-a-javascript-array
  const shuffled = array.toSorted(() => Math.random() - 0.5);
  return shuffled.slice(0, quantity);
}

// recycle the function above!
export function sample(array = []) {
  return sampleMany(array, 1)[0];
}

export function random(max, min) {
  return Math.floor(Math.random() * (max - min) + min);
}
```

Now the users from your fixture files can become just:

```js
import { random } from './utils.js';

export const users = defaultUsers.map((user) => ({
  ...user,
  credits: random(1000000, 500000),
  listings: Array.from({ length: random(15, 5) }, () => ({
    price: random(5000000, 200000),
  })),
}));
```

(`Array.from({ length: x }, () => ...)` is a quick way to get an array of length `x` in JS. the callback function works like in `.map()`)

Finally your `import.js` script:

```js
import { sampleMany, sample, random } from "./utils.js";

...

const createdStarships = await Starship.insertMany(starships);

...

await Promise.all(
  users.map(async ({ listings, ...user }, index) => {
    console.log(`Creating user: ${index}...`);
    const newUser = await User.create(user);

    console.log(`Saving listings for user: ${index}...`);

    const savedListings = await Listing.insertMany(
      listings.map((listing) => ({
        ...listing,
        user: newUser._id,
        starship: sample(createdStarships),
      }))
    );

    newUser.listings = savedListings.map((x) => x._id);
    newUser.starships = sampleMany(createdStarships, random(5, 1));

    await newUser.save();
  })
);
```

Final: make sure that your `listings` and `starships` are consistent - I assume you can list starships you don't own!

Little trick:

- starships in collection: sample x from the original full list;
- listings: sample n < x from the collection!

**Pull request merged**